### ISSUE TITLE
SBAT NNLS negative values

### ISSUE THREAD

**Issue by jerome-f** [rgcgithub/regenie#558]
@joellembatchou 

I noticed this error when using SBAT

```
-computing joint association tests...
ERROR: computing NNLS weight(i =2) & component(j = 4 out of 10) failed; pnorm(V11inv)*pnorm(V220) returned negative value (-8.15179e-07)
p_dm:
M=34
ERROR: computing NNLS weight(i =2) & component(j = 3 out of 10) failed; pnorm(V11inv)*pnorm(V220) returned negative value (-5.12633e-05)
p_mb:
M=33
ERROR: computing NNLS weight(i =4) & component(j = 3 out of 10) failed; pnorm(V11inv)*pnorm(V220) returned negative value (-0.00124637)
```

some of these are vanishingly small negative values and I wanted to ask if this might be an error due to `_eps_` correction (or lack thereof) in numerics. The masks this is happening at are very sparse and contain small number of variants.

**Comment by jerome-f** [comment]
I think the potential issues maybe in these code blocks 
```
Eigen::MatrixXd _inverse(const Eigen::MatrixXd& V)
{
  int n = V.rows();
  MatrixXd Vinv = V.llt().solve(MatrixXd::Identity(n, n));
  return(Vinv);
}
```
if the matrix is not positive semidefinite the this V.llt().solve() can have numerics issue I see here that it is being checked

```
Eigen::MatrixXd _npd(const Eigen::MatrixXd& X, double eps = 1e-6)
{
  // method 1: diag(X) = diag(X) + eps
  /* V = 0.5 * (V + V.transpose()); // force it symmetric */
  /* double eps = 1e-4; // to be a function argument */
  /* V.diagonal().array() += eps; // force positive-definite */
  
  // method 2: EVD (negative eigenvalues -> 0) + diag(X) = diag(X) + eps
  MatrixXd Y = 0.5 * (X + X.transpose());
  SelfAdjointEigenSolver<MatrixXd> solver(Y);
  VectorXd D = solver.eigenvalues();
  MatrixXd V = solver.eigenvectors();
  VectorXd Dplus = D.cwiseMax(0.0);
  MatrixXd Z = V * Dplus.asDiagonal() * V.transpose();
  Z.diagonal().array() += eps;

  return(Z);
}

```

should eps here be lower ?

**Comment by joellembatchou** [comment]
Hi Jerome,

While it shows up as an error message, the p-values are still computed for all the other tests and e.g. BURDEN-ACAT can be used to summarize burden signal.  We are aware of this error and , as you mentioned, this occurs quite rarely so we won't prioritize getting this to work for these edge cases.

Cheers,
Joelle

**Comment by jerome-f** [comment]
Hi Joelle, FWIW I reduced the `abseps` in NNLS.hpp to 1e-2, and don't observe this anymore. it is just one order of relaxation on the error tolerance for convergence. But is you are already using --sbat-adapt it should be okay. imo. I updated as few more functions so it is bit more robust.

**Comment by jerome-f** [comment]
@joellembatchou I have since performed a number of burden runs in qt traits and didn't observe this ERROR after relaxing the abseps to 10^-2. 

Note:not fixing this issue will affect all NNLS fits where the weight can be positive but the jburden_pnorm returns -1 or negative values to both components of the weight. The c wrapper in the external_lib returns -1 if abseps is not met, -2 and -3 for other errors. Also the fortran code is ~10 years old, the mvtnorm code base has since been updated to be more robust not critical for this specific issue but wanted to draw your attention to it.
